src/dtls_combine.rs

```diff
@@ -0,0 +1,219 @@
+    let ordered = Ordered::new(fragments);
+
+    // Investigate each fragment_offset + fragment_length to figure out
+    // the max contiguous range over the fragments.
+    let max = ordered.max_contiguous();
+
+    // Unwrap is OK, because we have at least one item (checked above).
+    let first_handshake = ordered.iter().next().unwrap();
+
+    if first_handshake.fragment_offset != 0 {
+        // The first fragment must start at 0, or we might have
+        // missing packets or arriving out of order.
+        return Ok((&[], None));
+    }
+
+    let msg_type = first_handshake.msg_type;
+    let message_seq = first_handshake.message_seq;
+    let length = first_handshake.length;
+
+    #[allow(clippy::comparison_chain)]
+    if max > length {
+        // Error case 4
+        return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+    } else if max < length {
+        // We do not have all fragments yet
+        return Ok((&[], None));
+    }
+
+    // Write the header into output.
+    {
+        out[0] = msg_type.into(); // The type.
+        out[1..4].copy_from_slice(&length.to_be_bytes()[1..]); // 24 bit length
+        out[4..6].copy_from_slice(&message_seq.to_be_bytes()); // 16 bit message sequence
+        out[6..9].copy_from_slice(&[0, 0, 0]); // 24 bit fragment_offset, which is 0 for the entire message.
+        out[9..12].copy_from_slice(&length.to_be_bytes()[1..]); // 24 bit fragment_length, which is entire length.
+    }
+
+    let data = &mut out[MESSAGE_HEADER_OFFSET..];
+
+    if data.len() < length as usize {
+        // Error case 1
+        return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+    }
+
+    // Loop the fragments, in order and output the data.
+    for handshake in ordered.iter() {
+        if msg_type != handshake.msg_type {
+            // Error case 2
+            return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+        }
+
+        if handshake.length != length {
+            // Error case 3
+            return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+        }
+
+        if handshake.message_seq != message_seq {
+            // Error case 6
+            return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+        }
+
+        let from = handshake.fragment_offset as usize;
+        let to = from + handshake.fragment_length as usize;
+
+        let body = match &handshake.body {
+            DTLSMessageHandshakeBody::Fragment(v) => v,
+            _ => {
+                // Error case 5
+                return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+            }
+        };
+
+        // Copy into output.
+        data[from..to].copy_from_slice(&body[..]);
+    }
+
+    // This parse should succeed now and produce a complete message.
+    let (rest, message) = parse_dtls_message_handshake(out)?;
+
+    Ok((rest, Some(message)))
+}
+
+struct Ordered<'a, 'b>([usize; MAX_FRAGMENTS], &'a [DTLSMessageHandshake<'b>]);
```

Comment:
I am missing the entire logic of this structure, this may require some more explanations.
As far as I understand, the field `0` is the sorted list of indices, the fragments being in the field `1`.

Why not create and use directly a list of sorted fragments ? Or, to better deal with lifetimes, references to fragments ?

You would just have to iterate the list of fragments, and insert them in a sorted structure (list, tree, etc).

Reply:
The goal here is to avoid an extra allocation. The user provides `&[DTLSMessageHandshake]`, which could have been received in any order (given this is UDP). The order we want is in the `fragment_offset` field of each message. Thus, the `Ordered` iterator iterates the messages without any extra heap allocation.

Personally I think this is better than allocating, but I should provide this motivation as code doc. However if you prefer new `Vec`, I can make that too.

Reply:
Thank you for clarifying. The methods seems globally fine, and there is no need to switch to an allocation.

That said, I think the code needs some comments to be more readable later. This is my main concern.
Also, the fields could be named (for ex `indices`/`fragments`) for clarity, and avoid `.0` and `.1`.